feat: add interact panel for interaction presentation

alexkornitzer opened the pull request:
This adds the UI/UX and plumbing need to present interactions to a player. This adds the interact panel along with its components such as menus and options. It also adds support for the examine interaction which has its own key binding and presentation. Due to this requirement the panel and friends has been designed in as generic a fashion as possible so that is can be used for all our interaction needs.


Fixes #119

alexkornitzer committed:
feat: add interact panel for interaction presentation


This adds the UI/UX and plumbing need to present interactions to a
player. This adds the interact panel along with its components such as
menus and options. It also adds support for the examine interaction
which has its own key binding and presentation. Due to this requirement
the panel and friends has been designed in as generic a fashion as
possible so that is can be used for all our interaction needs.


Fixes #119

alexkornitzer commented:
Work and progress will be tracked in #99, this is just here to track the code.

alexkornitzer commented:
Okay this is now ready for review, it adds the following:
- Interact Panel and friends used to show interactions, be that interacts or examines.
- Adds an examine input action but not wired
- Will layout the examine & interact panels around interactables
- Interacts are fully wired to support the multi interactions.

this-is-envy commented on Scenes/Components/Inventory/Item.gd:
```diff
@@ -21,6 +21,10 @@ extends Node2D
 			_collision.shape.radius = interactable_radius
 		update_configuration_warnings()
 
+var size: Vector2:
```
nit: s/size/uiPosition/?

it looks like this is returning location we should use to position the UI instead of a size?

this-is-envy replied:
Oh, the other thought I had here was that naming it something ui specific meant it would be less confusing if we decide to let interactables explicitly position their action toasts, e.g., large objects etcetc

alexkornitzer replied:
That was a typo good spot its meant to be size not position. The idea behind this is that we need to be able to get the size of NPC & Objects so that we can position around them. If we want to free form position I think that should be added when needed.

this-is-envy commented on Scripts/Components/NPC.gd:
```diff
@@ -16,6 +16,10 @@ extends Character
+		print(self.name)
+		if self.name == "mayor_buckley2":
+			print("adding another")
+			_talk_sensor.action_map[Enums.ActionVerb.PUSH_PULL] = []
```
rm; pretty sure this is debug cruft or validation cruft?

this-is-envy commented on Scripts/Components/Character.gd:
```diff
@@ -30,6 +30,17 @@ signal equipment_changed(id: String)
+		if sprite:
+			return sprite.get_rect().position
+		var asprite: AnimatedSprite2D = self.find_child("AnimatedSprite2D")
+		if asprite and asprite.sprite_frames.has_animation("default"):
```
maybe something like?
```
if asprite:
    frames = asprite.sprite_frames
    aname = frames.get_animation_names()
    if len(aname) > 0:
        animation = aname[0]
        return frames.get_frame_texture(animation, 0).get_size()
```

this-is-envy commented on Scripts/Components/CharacterControl/States/Interact.gd:
```diff
@@ -3,20 +3,53 @@ extends CharacterState
+		return
+	if _has_entered:
+		var interactable := _tgt.get_interactable()
+		if Input.is_action_pressed("interact"):
```
instead of using the default Input.is_action_pressed here let me introduce you to my past mistakes: [ControllerBase](https://github.com/Small-Loan-Studio/TGO/blob/tgo_dev/Scripts/Components/CharacterControl/ControllerBase.gd). It's an abstraction over Input that lets us specify custom controllers in the event we want to drive character behavior programmatically / without relying on input devices.

The tl;dr is basically instead of using the Input singleton a CharacterState [has a context](https://github.com/Small-Loan-Studio/TGO/blob/tgo_dev/Scripts/Components/CharacterControl/States/CharacterState.gd#L7) that contains a reference to the associated character's ControllerBase so instead of `Input.is_action_pressed` instead give this a shot:

```
if _ctx.controller.just_pressed(Enums.InputAction.INTERACT):
```

it _should_ be a 1:1 replacement. If it doesn't work make sure the button you're checking is registered [in controller setup](https://github.com/Small-Loan-Studio/TGO/blob/tgo_dev/Scripts/Components/Character.gd#L70-L87).

If it's not drop in file a bug and stick to Input but even then to using the enum->button name mapper, e.g.,

```
if Input.is_action_just_pressed(Enums.input_action_name(Enums.InputAction.UP)):
```

this-is-envy commented on Scripts/Components/CharacterControl/States/Interact.gd:
```diff
@@ -3,20 +3,53 @@ extends CharacterState
 func run_tick(_delta: float, change_state: Callable) -> void:
+	if _has_entered:
+		return
 	if _tgt.is_moveable_block():
```
I need to check off main bc I suspect this is is carry-over from the original push/pull rewrite but if you "ungrab" a block while moving it and pressing a different direction than "push" you'll swap to moving the block in that direction without ungrabbing.

alexkornitzer replied:
Not sure I understand this one, but let me get the grab/release showing up in the panel first.

this-is-envy replied:
oh, i just dropped this here as a reminder that i found a push/pull bug. I doubt it was from this state machine change but i hadn't seen it before so i need to test on main ... except it's a pain to change branches rn since I have a bunch of extremely in progress stuff for wwise. thus this note

this-is-envy commented on project.godot:
```diff
@@ -129,6 +129,11 @@ interact={
 }
+examine={
+"deadzone": 0.5,
+"events": [Object(InputEventJoypadButton,"resource_local_to_scene":false,"resource_name":"","device":-1,"button_index":2,"pressure":0.0,"pressed":false,"script":null)
```
let's get a keyboard mapping for this as well

this-is-envy commented on Scripts/Components/CharacterControl/States/Interact.gd:
```diff
@@ -3,20 +3,53 @@ extends CharacterState
 @export var idle_state: State
 @export var push_pull_state: State
 
+var _has_entered: bool = false
```
it looks like this this represents "has entered into the selection dialogue when multiple options are present"?

maybe rename to `_selecting_action` for clarity?

this-is-envy commented on Scripts/Components/Interactable.gd:
```diff
@@ -43,6 +57,50 @@ func _ready() -> void:
+
+	_interact_scene = InteractPanelScene.instantiate()
+	_interact_scene.actions.assign(action_map.keys())
+	_interact_scene.input = Enums.InputAction.INTERACT
```
Because we don't currently have any mappings for Examine let's
- only show examine when there is a corresponding action
- remove the EXAMINE action from the keyset when assigning to interact

```
var examine_actions: Array[Effect] = action_map[Enums.ActionVerb.EXAMINE]
var has_examine := len(examine_actions) > 0:

var other_actions: Array[Enums.ActionVerb] = action_map.keys().filter(
    func(e: Enums.ActionVerb)->bool: return e != Enums.ActionVerb.EXAMINE))
```

then we only instantiate examine & present if `has_examine` is true

this-is-envy commented on Scenes/Components/interacting/interact_panel.gd:
```diff
@@ -0,0 +1,196 @@
+		Enums.InputAction.EXAMINE:
+			return "Examine"
+		Enums.InputAction.INTERACT:
+			return "Interact"
```
might as well use `Enums.action_verb_name(EXAMINE|INTERACT)` here as well

this-is-envy commented on Scenes/Components/interacting/interact_menu.gd:
```diff
@@ -0,0 +1,46 @@
+		var option := InteractOptionScene.instantiate()
+		option.label = Enums.action_verb_name(action)
+		option.layout = layout
+		option.symbol = Image.load_from_file("res://Art/interacting/symbol_cross.png")
```
push the paths for the buttons up into an Enum or Util func whichever feels better. It'll be nice to start out with a single place to change the res paths when we get around to swapping out the PS symbols for Keyboard/Xbox etc.

this-is-envy commented:
Hey! This was a massive amount of work super hyped that you took it on 🙇.

I'm basically onboard with this and think that the shortcuts you took are all solid. There are a few bugs that you can either defer (my suggestion) or pick up and a handful of mostly nit-picky stuff that would be good to see land pre-merge.

## Bugs we should probably file for future work

- push/pull is still routing through the old path -- this looks intentional but I think that we do need to have it handled in the same way. I expect that we'll want to have some moveable objects that can be examined.

<img width="277" alt="Screen Shot 2025-03-30 at 3 01 39 PM" src="https://github.com/user-attachments/assets/b4aef6cf-42e5-496a-a77c-40f8dc82cd10" />

- Because we add these as controls/scenes as a child of the interactable we're getting fucked by draw order. This _might_ be easy to fix by setting z-index or might be tricky if you have to jump up to the canvas layer bc then you're disconnected from the relevant objects positions

<img width="408" alt="Screen Shot 2025-03-30 at 2 57 45 PM" src="https://github.com/user-attachments/assets/a2faa3bd-4b06-4a63-8216-a6a00d240e8a" />


- Similarly because the toast is a "normal" scene we get shadowed in the dark. That might actually be preferred but idk

<img width="453" alt="Screen Shot 2025-03-30 at 3 04 19 PM" src="https://github.com/user-attachments/assets/d522d4d3-e389-4eb6-b158-9187bc82cce8" />

alexkornitzer commented:
Consulted the bossman on the z-index stuff and he for now wants it to show cleanly, so I will try whacking the z-index up and if that is not working i'll change it to add into the HUD layer.

Otherwise will work through your suggestions above and then we can work out what needs new issues tickets like you said.

alexkornitzer committed:
feat: allow the panel to be placed globally


This updates the panel so that it can be placed globally in the world,
this works by adding a presentation canvas layer to the world. We can't
directly add to the world or the tween of the day/night cycle will tween
everything. So a new canvas layer is added that follows the viewport.

alexkornitzer commented:
Everything is done except for the showing a panel on movable block cause that needs further discussion.

this-is-envy commented:
amaze. going to go ahead and merge this so it hits sprint3; we can confer on any stuff that needs to get ticketed etc whetever
